main_loop_tf/main.py

```diff
@@ -634,22 +722,53 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
 
         if is_training:
             train_summary_op = tf.summary.merge(summaries['training'])
+            # Merge only the training summaries of the gpus used.
+            # This is done searching by name since the variables are
+            # always preceded by 'Tower{0,1,2,..}' or 'GPU{0,1,2,..}'
+            training_summaries = []
+            for i in range(smaller_n_gpus):
+                for el in summaries['training']:
+                    if 'GPU%d' % i in el.name or 'Tower%d' % i in el.name:
+                        training_summaries += [el]
+            s_train_summary_op = tf.summary.merge(training_summaries)
         else:
             val_summary_ops = {}
+            s_val_summary_ops = {}
             for k, s in summaries.iteritems():
+                # Same reasoning done for the training summaries (see
+                # above)
+                val_summaries = []
+                for i in range(smaller_n_gpus):
+                    for el in s:
+                        if 'GPU%d' % i in el.name or 'Tower%d' % i in el.name:
+                            val_summaries += [el]
                 val_summary_ops[k] = tf.summary.merge(s)
-
+                s_val_summary_ops[k] = tf.summary.merge(val_summaries)
     if is_training:
-        return [avg_tower_loss, train_op], train_summary_op, reset_cm_op
+        return ([[avg_tower_loss, train_op], [s_avg_tower_loss, s_train_op]],
+                [train_summary_op, s_train_summary_op],
+                [reset_cm_op, s_reset_cm_op])
     else:
-        return ([preds, softmax_preds, m_iou, per_class_iou, avg_tower_loss,
-                 cm_update_op], val_summary_ops, reset_cm_op)
+        return ([[preds, softmax_preds, m_iou, per_class_iou, avg_tower_loss,
+                 cm_update_op], [s_preds, s_softmax_preds, s_m_iou,
+                                 s_per_class_iou, s_avg_tower_loss,
+                                 s_cm_update_op]],
+                [val_summary_ops, s_val_summary_ops], [reset_cm_op,
+                                                       s_reset_cm_op])
 
 
 def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
               val_outs, val_summary_ops, val_reset_cm_op, loss_fn, Dataset,
-              dataset_params, valid_params, sv, saver):
+              dataset_params, valid_params, sv, saver, smaller_n_gpus):
+
+    inputs_per_gpu, labels_per_gpu, prev_err = placeholders
 
+    # Two separate list of placeholders according to the number of gpus
+    # used (p_list is used when all gpus are used while the second in
+    # the case of smaller batch)
+    p_list = inputs_per_gpu + labels_per_gpu + [prev_err]
+    smaller_p_list = (inputs_per_gpu[:smaller_n_gpus['train']] +
+                      labels_per_gpu[:smaller_n_gpus['train']] + [prev_err])
```

Comment:
Rather than saving two p_lists, why don't you create `p_list` in the main loop depending on the size of the batch (i.e., either equal to the full sum or the sum of the subsets) and just use it as is later on?